File: typescript/12_inheritance.ts
// ============================================================
// Topic:  Inheritance
// Summary: A subclass reuses a base class, overrides methods, calls super.
// Run:     cd typescript && npx ts-node 12_inheritance.ts
// ============================================================

class Animal {
  constructor(public name: string) {}

  sound(): string {
    return "some sound";
  }

  describe(): string {
    return `${this.name} says ${this.sound()}`;
  }
}

// ----- 1. extends + super() to reuse the base -----
class Dog extends Animal {
  constructor(name: string, public breed: string) {
    super(name); // run Animal's constructor
  }

  sound(): string {
    return "Woof"; // override
  }

  describe(): string {
    return `${super.describe()} (${this.breed})`; // extend base behavior
  }
}

// ----- 2. Subclass that only overrides one method -----
class Cat extends Animal {
  sound(): string {
    return "Meow";
  }
}

console.log(new Animal("Creature").describe());     // Output: Creature says some sound
console.log(new Dog("Rex", "Labrador").describe()); // Output: Rex says Woof (Labrador)
console.log(new Cat("Whiskers").describe());        // Output: Whiskers says Meow

// Cross-language note:
//   TS: `class Dog extends Animal`; `super(...)` calls the base ctor, `super.m()` the base method.
//   Methods are virtual by default — describe() (on Animal) calls this.sound() and
//   dispatches to the subclass override. TS has single inheritance (one base class).

File: python/12_inheritance.py
# ============================================================
# Topic:  Inheritance
# Summary: A subclass reuses a base class, overrides methods, calls super().
# Run:     python python/12_inheritance.py
# ============================================================

class Animal:
    def __init__(self, name):
        self.name = name

    def sound(self):
        return "some sound"

    def describe(self):
        return f"{self.name} says {self.sound()}"


# ----- 1. Subclass with extra state; super() calls the base constructor -----
class Dog(Animal):
    def __init__(self, name, breed):
        super().__init__(name)      # run Animal.__init__
        self.breed = breed

    def sound(self):                # override
        return "Woof"

    def describe(self):             # extend the base behavior
        return f"{super().describe()} ({self.breed})"


# ----- 2. Subclass that only overrides one method -----
class Cat(Animal):
    def sound(self):
        return "Meow"


print(Animal("Creature").describe())     # Output: Creature says some sound
print(Dog("Rex", "Labrador").describe()) # Output: Rex says Woof (Labrador)
print(Cat("Whiskers").describe())        # Output: Whiskers says Meow

# Cross-language note:
#   Python: class Dog(Animal) inherits; super() reaches the base ctor/method.
#   describe() (defined once on Animal) calls self.sound(), which dispatches to the
#   subclass override — that's runtime polymorphism (see 13_polymorphism).
